metronome/core/src/io/iohk/metronome/core/messages/RPCTracker.scala

```diff
@@ -0,0 +1,102 @@
+import scala.reflect.ClassTag
+
+/** `RPCTracker` can be used to register outgoing requests and later
+  * match them up with incoming responses, thus it facilitates turning
+  * the two independent messages into a `Kleisli[F, Request, Option[Response]]`,
+  * by a component that has access to the network, where a `None` result means
+  * the operation timed out before a response was received.
+  *
+  * The workflow is:
+  * 0. Receive some request parameters in a method.
+  * 1. Create a request ID.
+  * 2. Create a request with the ID.
+  * 3. Register the request with the tracker, hold on to the handle.
+  * 4. Send the request over the network.
+  * 5. Wait on the handle, eventually returning the optional result to the caller.
+  * 6. Pass every response received from the network to the tracker (on the network handler fiber).
+  */
+class RPCTracker[F[_]: Timer: Concurrent, M](
+    deferredMapRef: Ref[F, Map[UUID, RPCTracker.Entry[F, _]]],
+    defaultTimeout: FiniteDuration
+) {
+  import RPCTracker.Entry
+
+  def register[
+      Req <: RPCMessageCompanion#Request,
+      Res <: RPCMessageCompanion#Response
+  ](
+      request: Req,
+      timeout: FiniteDuration = defaultTimeout
+  )(implicit
+      ev1: Req <:< M,
+      ev2: RPCPair.Aux[Req, Res],
+      // Used by `RPCTracker.Entry.complete` to make sure only the
+      // expected response type can complete a request.
+      ct: ClassTag[Res]
+  ): F[F[Option[Res]]] = {
+    val requestId = request.requestId
+    for {
+      d <- Deferred[F, Option[Res]]
+      e = RPCTracker.Entry(d)
+      _ <- deferredMapRef.update(_ + (requestId -> e))
+      _ <- Concurrent[F].start {
+        Timer[F].sleep(timeout) >> completeWithTimeout(requestId)
+      }
+    } yield d.get
+  }
+
+  def complete[Res <: RPCMessageCompanion#Response](
+      response: Res
+  )(implicit ev: Res <:< M): F[Boolean] = {
+    remove(response.requestId).flatMap {
+      case None    => false.pure[F]
+      case Some(e) => e.complete(response)
+    }
+  }
+
+  private def completeWithTimeout(requestId: UUID): F[Unit] =
+    remove(requestId).flatMap {
+      case None    => ().pure[F]
+      case Some(e) => e.timeout
+    }
+
+  private def remove(requestId: UUID): F[Option[Entry[F, _]]] =
+    deferredMapRef.modify { dm =>
+      (dm - requestId, dm.get(requestId))
+    }
+}
+
+object RPCTracker {
+  case class Entry[F[_]: Sync, Res](
+      deferred: Deferred[F, Option[Res]]
+  )(implicit ct: ClassTag[Res]) {
+    def timeout: F[Unit] =
+      deferred.complete(None).attempt.void
+
+    def complete[M](response: M): F[Boolean] = {
+      response match {
+        case expected: Res =>
+          deferred.complete(Some(expected)).attempt.map(_.isRight)
+        case _ =>
+          // Wrong type, as evidenced by `ct` not maching `Res`.
+          // Not returning an error because if the message arrived after
+          // the timeout we wouldn't know anyway what it was meant to complete.
```

Comment:
Perhaps the error wouldn't be that useful, but I don't quite get the argumentation. This code path isn't run when a timeout occurs so we would be able to differentiate those cases.

Reply:
Right, that wasn't very clear. What I meant was that if a timeout occurred independently and earlier than `tracker.complete(wrongResponse)` is called, then the `Entry` would be gone by this time and we couldn't differentiate why the response hasn't been handled, it would look like any other timeout. 

But you may be right, it could be worth differentiating anyway, just to catch these problems earlier.

Reply:
Added different handling for errors due to type mismatch.